### PR TITLE
Packaging: Virtual package for libllvm

### PR DESCRIPTION
**Pull request by kwryankrattiger**
Add an abstraction around libllvm to allow libllvm
providers to be specified for all packages.

This is targeting allowing mesa to build against
llvm-amdgpu or intel-llvm or llvm or any other
custom llvm variant that arises for specific GPU
toolchains.

@chuckatkins @srekolam

### CODE REVIEW

**Comment by trws** [comment]
What benefit does this have for mesa built this way? Adding a virtual adds some non-trivial  issues, especially in terms of variant propagation.  Is the cost worth it?

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py [hunk 1]
```diff
@@ -34,6 +34,17 @@ class LlvmAmdgpu(CMakePackage):
...
+        "llvm_dylib",
+        default=False,
+        description="Build LLVM shared library, containing all "
+        "components in a single shared library",
+    )
+
+    # TODO: Version mapping needs more detail
+    provides('libllvm@:12', when='@3.5:4.2')
+    provides('libllvm@12', when='@4.2:4.3')
+    provides('libllvm@13', when='@4.3:')
```
Comment (trws) [inline var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: Does it provide the appropriate library when llvm_dylib is False?

Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: Oops, I messed up the patch on this, it used to have a
```python
provides('libllvm +llvm_dylib', when='+llvm_dylib')
```

**Comment by kwryankrattiger** [comment]
> What benefit does this have for mesa built this way? Adding a virtual adds some non-trivial issues, especially in terms of variant propagation. Is the cost worth it?

Currently when we build the rocm stack, it is required to build `mesa  ~llvm` because it requires it's own libllvm. However, some of the `mesa +llvm` extensions are required down stream for other packages. By doing it this way, mesa can depend on the `libllvm` virtual package when `+llvm`, and build with the required extensions for downstream packages.

**Comment by trws** [comment]
Let me rephrase, is this to produce a better mesa in some fashion, or to avoid building two copies of LLVM? It still might be worth it, but having disparate LLVM packages is going to get a great deal more painful if we end up with many things like this.
________________________________

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py (hunk 1 above)
Comment (srekolam) [inline var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: provides('libllvm@11', when='@3.5:3.8')   
provides('libllvm@12', when='@3.9:4.2') 
provides('libllvm@13', when='@4.3:')

**Review by trws — DISMISSED** [review]
Ok, I'm actually good with this at this point.  I'll hold back on merge to give @tgamblin a chance to take a look but I don't see any issues with the package updates at this point.

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 2]
```diff
@@ -149,6 +149,19 @@ class Llvm(CMakePackage, CudaPackage):
...
+    provides('libllvm@12', when='@12:12')
+    provides('libllvm@11', when='@11:11')
+    provides('libllvm@10', when='@10:10')
+    provides('libllvm@9', when='@9:9')
+    provides('libllvm@8', when='@8:8')
+    provides('libllvm@7', when='@7:7')
+    provides('libllvm@6', when='@6:6')
+    provides('libllvm@5', when='@5:5')
+    provides('libllvm@4', when='@4:4')
+    provides('libllvm@3', when='@3:3')
```
Comment (haampie) [inline var/spack/repos/builtin/packages/llvm/package.py]:
```
In [1]: spack.spec.Spec('@11:11')
Out[1]: @11
```
I'm afraid these ranges are currently buggy. To ensure they are actually ranges you have to use `11.0.0:11` now.

**Comment by haampie** [comment]
I'm not sure I agree with the versions for libllvm, they are rather meaningless. The problem is that e.g. llvm-amdgpu may say it's version 13, but for all we know it's 3 commits ahead of upstream llvm 12 + a bunch of commits that aren't upstreamed. So, after this, does it really make sense to talk about libllvm@12/13?

**Comment by kwryankrattiger** [comment]
> I'm not sure I agree with the versions for libllvm, they are rather meaningless. The problem is that e.g. llvm-amdgpu may say it's version 13, but for all we know it's 3 commits ahead of upstream llvm 12 + a bunch of commits that aren't upstreamed. So, after this, does it really make sense to talk about libllvm@12/13?

I agree that the mapping is kind of arbitrary. The main thing that I am trying to capture is API compatibility. My assumptions, be they either good or bad, are
 1. that llvm is not breaking API compatibility within a major version
 2. llvm-amdgpu (and others) do not break API compatibility from the major version they are loosely based on

File: var/spack/repos/builtin/packages/llvm/package.py (hunk 2 above)
Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/llvm/package.py]: ah, I had seen talk about this on slack and was hoping it would just be that easy. I will update those!

Comment (haampie) [reply var/spack/repos/builtin/packages/llvm/package.py]: It is just x:x which is not working so well :D

**Comment by kwryankrattiger** [comment]
@haampie Any more comments on the version mapping? Does the approach seem reasonable or is there possibly a better approach we should explore.

@tgamblin Did you still want to review this?

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py [hunk 3]
```diff
@@ -34,6 +34,17 @@ class LlvmAmdgpu(CMakePackage):
...
+        "llvm_dylib",
+        default=False,
+        description="Build LLVM shared library, containing all "
+        "components in a single shared library",
+    )
+
+    provides('libllvm@11', when='@3.5:3.8.0')
+    provides('libllvm@12', when='@3.9:4.2.0')
+    provides('libllvm@13', when='@4.3:')
+    provides('libllvm +llvm_dylib', when='+llvm_dylib')
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: Unfortunately I'm pretty sure the variant here on the `libllvm` virtual has no effect (our fault, but we can fix it in the new concretizer).  Do you need it or is `libllvm` sufficient?  It seems like you're using `libllvm` alone to mean that `llvm_dylib` is enabled on the provider.

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 4]
```diff
@@ -156,6 +156,19 @@ class Llvm(CMakePackage, CudaPackage):
...
+    provides('libllvm@11', when='@11.0.0:11')
+    provides('libllvm@10', when='@10.0.0:10')
+    provides('libllvm@9', when='@9.0.0:9')
+    provides('libllvm@8', when='@8.0.0:8')
+    provides('libllvm@7', when='@7.0.0:7')
+    provides('libllvm@6', when='@6.0.0:6')
+    provides('libllvm@5', when='@5.0.0:5')
+    provides('libllvm@4', when='@4.0.0:4')
+    provides('libllvm@3', when='@3.0.0:3')
+    provides('libllvm +llvm_dylib', when='+llvm_dylib')
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/llvm/package.py]: same here.

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py [hunk 5]
```diff
@@ -34,6 +34,17 @@ class LlvmAmdgpu(CMakePackage):
...
     variant('rocm-device-libs', default=True, description='Build ROCm device libs as external LLVM project instead of a standalone spack package.')
     variant('openmp', default=True, description='Enable OpenMP')
+    variant(
+        "llvm_dylib",
+        default=False,
+        description="Build LLVM shared library, containing all "
+        "components in a single shared library",
+    )
+
+    provides('libllvm@11', when='@3.5:3.8.0')
```
Comment (kwryankrattiger) [inline var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: drop `0` from right side version

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py (hunk 3 above)
Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: I thought I would need to map the variants to the virtual package, but if that is done automatically then that is sufficient for this case. The idea here was that `libllvm ~llvm_dylib` was the default and then if `libllvm +llvm_dylib` was requested it would provide `llvm-amdgpu +llvm_dylib`.

**Comment by kwryankrattiger** [comment]
Rebased on develop.

**Comment by chuckatkins** [comment]
> I'm not sure I agree with the versions for libllvm, they are rather meaningless. The problem is that e.g. llvm-amdgpu may say it's version 13, but for all we know it's 3 commits ahead of upstream llvm 12 + a bunch of commits that aren't upstreamed. So, after this, does it really make sense to talk about libllvm@12/13?

We're addressing this upstream to be able test in the code for whether or not a pre-release is being used which will fix issues with how `llvm-amdgpu` is being used in mesa.
https://reviews.llvm.org/D115818

**Comment by chuckatkins** [comment]
@tgamblin I believe this is ready to go with all requested changes being addressed.  Is there anything holding this up to getting it merged?

**Review by tgamblin — APPROVED** [review]
